### PR TITLE
[FLINK-23165][python] Add StreamExecutionEnvironment#registerSlotSharingGroup to PyFlink and Scala

### PR DESCRIPTION
**Pull request by SteNicholas**
## What is the purpose of the change

*Flink supports configuring fine grained resource requirements via DataStream API. PyFlink should add `StreamExecutionEnvironment#registerSlotSharingGroup`.*

## Brief change log

  - *Introduce `SlotSharingGroup` in pyflink. Users can use this class to describe the name and the the different resource components of a slot sharing group.*
  - *Add `DataStream#slotSharingGroup(SlotSharingGroup)`.*
  - *Add `StreamExecutionEnvironment#registerSlotSharingGroup(SlotSharingGroup)`.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-python/pyflink/datastream/tests/test_stream_execution_environment_completeness.py [hunk 1]
```diff
@@ -49,7 +49,7 @@ def excluded_methods(cls):
                 'createInput', 'createLocalEnvironmentWithWebUI', 'fromCollection',
                 'socketTextStream', 'initializeContextEnvironment', 'readTextFile',
                 'setNumberOfExecutionRetries', 'configure', 'executeAsync', 'registerJobListener',
-                'clearJobListeners', 'getJobListeners', "fromSequence", "registerSlotSharingGroup"}
+                'clearJobListeners', 'getJobListeners', "fromSequence"}
```
Comment (KarmaGYZ) [inline flink-python/pyflink/datastream/tests/test_stream_execution_environment_completeness.py:52]: Also, delete the TODO above.

**Review by KarmaGYZ — COMMENTED** [review]
Thanks for the PR, @SteNicholas . The logic generally looks good to me. We need someone who familiar with PyFlink to give another check. 
cc @HuangXingBo .

**Comment by KarmaGYZ** [comment]
The testing fails in format checking. Would you like to fix it?

**Comment by SteNicholas** [comment]
> The testing fails in format checking. Would you like to fix it?

I have fixed the checkstyle of the `slot_sharing_group.py`.

**Comment by KarmaGYZ** [comment]
Just out of curiosity, is there any tool that can automatically reformat the python code?

**Comment by SteNicholas** [comment]
> Just out of curiosity, is there any tool that can automatically reformat the python code?

No automatically reformat tool for PyFlink like Maven Spotless. Only use `lint-python.sh` to validate the checkstyle and the tests.

**Comment by KarmaGYZ** [comment]
Maybe we can complete the missing scala interface as well in this PR, with another commit. WDYT?

**Comment by SteNicholas** [comment]
> Maybe we can complete the missing scala interface as well in this PR, with another commit. WDYT?

@KarmaGYZ IMO, I would like to add scala interface into another commit in this PR.

File: flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/DataStream.scala [hunk 2]
```diff
@@ -327,6 +327,29 @@ class DataStream[T](stream: JavaStream[T]) {
     this
   }
 
+  /**
```
Comment (KarmaGYZ) [inline flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/DataStream.scala:330]: It's better to be split into another commit.

File: flink-python/pyflink/datastream/slot_sharing_group.py [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+    """
+    Describe the name and the the different resource components of a slot sharing group.
+    """
+
+    def __init__(self, j_slot_sharing_group):
+        self._j_slot_sharing_group = j_slot_sharing_group
+
+    def get_name(self) -> str:
+        """
+        Get the name of this SlotSharingGroup.
```
Comment (HuangXingBo) [inline flink-python/pyflink/datastream/slot_sharing_group.py:106]:
```suggestion
        Gets the name of this SlotSharingGroup.
```

File: flink-python/pyflink/datastream/slot_sharing_group.py [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+
+        :return: The name of the SlotSharingGroup.
+        """
+        return self._j_slot_sharing_group.getName()
+
+    def get_managed_memory(self) -> Optional[MemorySize]:
+        """
+        Get the task managed memory for this SlotSharingGroup.
```
Comment (HuangXingBo) [inline flink-python/pyflink/datastream/slot_sharing_group.py:114]: ditto

File: flink-python/pyflink/datastream/slot_sharing_group.py [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+
+        :return: The task managed memory of the SlotSharingGroup.
+        """
+        managed_memory = self._j_slot_sharing_group.getManagedMemory()
+        return MemorySize(managed_memory.get()) if managed_memory.isPresent() else None
+
+    def get_task_heap_memory(self) -> Optional[MemorySize]:
+        """
+        Get the task heap memory for this SlotSharingGroup.
```
Comment (HuangXingBo) [inline flink-python/pyflink/datastream/slot_sharing_group.py:123]: ditto

File: flink-python/pyflink/datastream/slot_sharing_group.py [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+        """
+        Builder for the SlotSharingGroup.
+        """
+
+        def __init__(self, j_builder):
+            self._j_builder = j_builder
+
+        def set_cpu_cores(self, cpu_cores: float) -> 'SlotSharingGroup.Builder':
+            """
+            Set the CPU cores for this SlotSharingGroup.
```
Comment (HuangXingBo) [inline flink-python/pyflink/datastream/slot_sharing_group.py:193]:
```suggestion
            Sets the CPU cores for this SlotSharingGroup.
```

**Review by HuangXingBo — COMMENTED** [review]
@SteNicholas Thanks a lot for the PR. Looks good overall. I only left a minor comments.

File: flink-python/pyflink/datastream/slot_sharing_group.py (hunk 3 above)
Comment (KarmaGYZ) [reply flink-python/pyflink/datastream/slot_sharing_group.py:106]: May I ask why we use the singular here? Isn't it an impressive mood? Also search in Linggle, the "Get the" is used more frequently. [1][2]

[1] https://linggle.com/?q=Get+the
[2] https://linggle.com/?q=Gets+the

**Comment by SteNicholas** [comment]
> @SteNicholas Thanks a lot for the PR. Looks good overall. I only left a minor comments.

@HuangXingBo , thanks for detailed review. I have addressed the comments you left. Please help to review again.

**Review by HuangXingBo — APPROVED** [review]
LGTM

**Comment by HuangXingBo** [comment]
> Just out of curiosity, is there any tool that can automatically reformat the python code?

**Comment by HuangXingBo** [comment]
@flinkbot run azure
